rascaline/src/calculators/lode/spherical_expansion.rs

```diff
@@ -327,24 +432,201 @@ impl CalculatorBase for LodeSphericalExpansion {
 
     #[time_graph::instrument(name = "LodeSphericalExpansion::compute")]
     fn compute(&mut self, systems: &mut [Box<dyn System>], descriptor: &mut TensorMap) -> Result<(), Error> {
-        let do_gradients = false;
 
-        // TODO: let the user provide the cutoff?
-        // TODO: check atomic_gaussian_width so that we have a non-zero number of k-vectors
-        let k_cutoff = 1.2 * std::f64::consts::PI / self.parameters.atomic_gaussian_width;
+        let k_cutoff = self.parameters.get_k_cutoff();
 
         for (system_i, system) in systems.iter_mut().enumerate() {
+            let species = system.species()?;
             let cell = system.cell()?;
             if cell.shape() == UnitCell::infinite().shape() {
                 return Err(Error::InvalidParameter("LODE can only be used with periodic systems".into()));
             }
 
             let k_vectors = compute_k_vectors(&cell, k_cutoff);
+            if k_vectors.is_empty() {
+                return Err(Error::InvalidParameter("No k-vectors for current combination of hyper parameters.".into()));
+            }
+
             assert!(!k_vectors.is_empty());
 
-            self.project_k_to_nlm(&k_vectors, do_gradients);
+            self.project_k_to_nlm(&k_vectors);
+
+            let structure_factors = compute_structure_factors(
+                system.positions()?,
+                system.species()?,
+                &k_vectors
+            );
+
+            let density_fourrier = self.compute_density_fourrier(&k_vectors);
+
+            let global_factor = 4.0 * std::f64::consts::PI / cell.volume();
+
+            // Add k = 0 contributions for (m, l) = (0, 0)
+            if self.parameters.potential_exponent == 0 || self.parameters.potential_exponent >= 4 {
+                let k0_contrib = &self.compute_k0_contributions();
+                for &species_neighbor in species {
+                    for center_i in 0..system.size()? {
+                        let block_i = descriptor.keys().position(&[
+                            0.into(),
+                            species[center_i].into(),
+                            species_neighbor.into(),
+                        ]).expect("missing block");
+
+                        let mut block = descriptor.block_mut(block_i);
+                        let values = block.values_mut();
+                        let array = values.data.as_array_mut();
+
+                        let sample = [system_i.into(), center_i.into()];
+                        let sample_i = match values.samples.position(&sample) {
+                            Some(s) => s,
+                            None => continue
+                        };
+
+                        for (_property_i, [n]) in values.properties.iter_fixed_size().enumerate() {
+                            let n = n.usize();
+                            array[[sample_i, 0, _property_i]] += global_factor * k0_contrib[[n]];
```

Comment:
It seems that this is not saved in the final feature array. When I compare the final features to pyLODE they only contain the 'main feature' and not the k_0 contribution. @Luthaf probably knows what's going on here.